### ISSUE TITLE
Understanding Clean Code Principles

### ISSUE THREAD

**Issue by 3thancs** [3thancs/3thancs-intern-repo#40]
#### 🎯 Goal  
Understand the core principles of clean code and why they matter in real-world development.  

---

### ✅ Tasks  
- [x] **Research and summarize the following clean code principles in `clean_code.md`:**  
  - Simplicity – Keep code as simple as possible.  
  - Readability – Code should be easy to understand.  
  - Maintainability – Future developers (including you!) should be able to work with the code easily.  
  - Consistency – Follow style guides and project conventions.  
  - Efficiency – Write performant, optimized code without premature over-engineering.  
- [x] **Find an example of messy code online (or write one yourself) and describe why it's difficult to read.**  
- [x] **Rewrite the code in a cleaner, more structured way.**  
- [x] **Commit and push your changes to GitHub.**  

---

**Comment by 3thancs** [comment]
### **Research and summarize the following clean code principles**
**Simplicity - Keep code as simple as possible**
The code logic should be straight forward, small and focused functions. Large, complex code can be difficult to modify, debug and test. Instead, large problems should be broken down into small, simple steps

**Readability - The code should be easy to understand**
Code ensures that readers are able to digest and understand the intention and behavior of the code. It is important since code is read more than writing and reduces misinterpretation. The code comments should explain the 'why' and abbreviation should be avoided (such as usr = user)

**Maintainability – Future developers (including you!) should be able to work with the code easily.**
The code should be easily modified to address bugs, add features or improve performance without regressing the system. Maintainability constitutes for most of the software lifecycle. Unmaintainable code can result in code rewrites. Code blocks should not be highly dependent to ensure high maintainability and code should be refactored until duplicate code or poor structure materializes.

**Consistency – Follow style guides and project conventions**
Consistency ensures that the code writing is consistent with the system - in terms of error handling, architectural patterns (such as object-oriented programming design patterns) and indentations. An automated formatter such as Prettier and ESLint ensures that indentations are addressed and code errors are underscored respectively.

**Efficiency – Write performant, optimized code without premature over-engineering**
Defines as writing code on optimizing based on proven bottlenecks - not on blind guesses of future performance needs. Over-optimized code results in complex, fragile and highly dependent code blocks. Under-optimized code results in performance slowdown. The moderate balance results in fast and clean code. Code blocks should be profiled, efficient data structures should be practiced, and write straightforward code first, before refactoring for performance if needed.


### **Example of messy code and why it is difficult to read**
<img width="1554" height="659" alt="Image" src="https://github.com/user-attachments/assets/2c65672f-b135-4831-8955-a84427482613" />
**Image above shows the messy code**

* **Why it is difficult to read**
  * Lack of indentation of between each line
  * Lack of code comments to denote the 'why' for certain code blocks or line. For example, the selection of dependent and independent variables were not written as code comments. This results in difficulty in debugging code blocks in isolation.
  * Lack of an explicit mention of x1, x2, y1 and y2 used for either training set or test set which can be difficult to trace back for debugging


**Rewrite the code in a cleaner, more structured way.**
<img width="859" height="468" alt="Image" src="https://github.com/user-attachments/assets/aeae3a18-d019-40a1-9fa8-294b5d785161" />
<img width="871" height="710" alt="Image" src="https://github.com/user-attachments/assets/b1c7cf03-f423-41d0-a02c-6e65168be945" />
<img width="645" height="322" alt="Image" src="https://github.com/user-attachments/assets/774eea1a-c182-408d-a3ae-ac12d7d79c31" />
<img width="865" height="584" alt="Image" src="https://github.com/user-attachments/assets/bc4bf8c0-5bc7-4a6e-8f2e-41e4f25e4666" />

**Comment by FocusBearQA** [comment]
✅ **Approved**

Great job! Here is my review of your submission:

- **Issue requires code and markdown reflection:** You have provided both a well-written summary of clean code principles in markdown and addressed the example code task.
- **Markdown file ('clean_code.md'):** You summarized each clean code principle clearly and with your own wording, showing genuine understanding. The explanations are easy to read and not just a copy of the prompt. Good use of formatting and breakdowns.
- **Messy code example & rewrite:** You included images showing the messy code, described clearly why it's difficult to read, and provided a cleaned-up version. Your justification for the improvements is specific.
- **Code quality:** Hard to fully assess since only screenshots were provided, but your explanations show you understand what the improvements were and why they matter.
- **Personal reflection:** Your points about readability and maintainability were personal and genuine (i.e., not just definitions), so it is credible that you did the task yourself.
- **Grammar/Spelling:** There are a few minor grammar issues (e.g., 'straight forward' should be 'straightforward', and some slightly awkward sentences), but nothing that impedes understanding.
- **AI-Generated Check:** Your submission does not appear to be AI-generated; it is tailored and shows personal engagement.

**Suggestions:**
- Consider fixing a couple of small grammar slips in future submissions.
- Whenever possible, it's better to include code snippets directly in markdown instead of screenshots to make it easier to read/review (unless screenshots are required to capture context).

Reviewed files: clean_code.md (and referenced images).

Keep up the good work! 🎉

🤖 Intern Review Bot: Approved
